backend/utils.py

```diff
@@ -1,25 +1,101 @@
+"""Backend utilities"""
 import json
+import operator
+from datetime import datetime
+from functools import reduce
+from typing import Any, Dict, List, Tuple, Union
 
 
 def cnt(vals):
-  return len(set(vals))
+    """Count values"""
+    return len(set(vals))
 
 
 def commify(n):
-  return f'{n:,}'
+    """Those dirty commies"""
+    return f'{n:,}'
 
 
 def dump(o):
-  return json.dumps(o, indent=2)
+    """Return pretty printed json"""
+    return json.dumps(o, indent=2)
 
 
 def pdump(o):
-  print(dump(o))
+    """Prent pretty printed json"""
+    print(dump(o))
 
 
-class Bunch(object):    # dictionary to namespace, a la https://stackoverflow.com/a/2597440/1368860
-  def __init__(self, adict):
-    self.__dict__.update(adict)
+class Bunch(object):
+    """dictionary to namespace, a la https://stackoverflow.com/a/2597440/1368860"""
+    def __init__(self, adict):
+        """Init"""
+        self.__dict__.update(adict)
 
-  def to_dict(self):
-    return self.__dict__
+    def to_dict(self):
+        """Convert to dictionary"""
+        return self.__dict__
+
+
+def get_nested_from_dict(d: Dict, key_path: List):
+    """Get nested value from dictionary"""
+    return reduce(operator.getitem, key_path, d)
+
+
+def set_nested_in_dict(d: Dict, key_path: List, value: Any):
+    """Set nested value in dictionary"""
+    # noinspection PyUnresolvedReferences
+    get_nested_from_dict(d, key_path[:-1])[key_path[-1]] = value
+
+
+def hierarchify_list_of_parent_kids(
```

Comment:
@Sigfried : `hierarchify_list_of_parent_kids()`: This is the function that does the recursion. I tried a bunch of different ways, but wasn't having trouble with duplicate subtrees, recursion depth, as well as trying to debug. I found this to be the easiest way, and it seems decently performant.

Reply:
@joeflack4: it's not working for me. 

http://127.0.0.1:8000/hierarchy?codeset_id=411456218|40061425|484619125|419757429 :
`Time to get concept_relationship_subsumes_only: 0:00:01.422682
Time to traverse: 0:00:00.765243
INFO:     127.0.0.1:53503 - "PUT /datasets/csets HTTP/1.1" 500 Internal Server Error
ERROR:    Exception in ASGI application
[stack trace is basically all FastAPI stuff]
`

What it's trying to return is:
`d
{(411456218, 138056): {}, (411456218, 252341): {}, (411456218, 252658): {}, (411456218, 252942): {}, (411456218, 254980): {}, (411456218, 255841): {}, (411456218, 256448): {}, (411456218, 256716): {}, (411456218, 257581): {}, (411456218, 257583): {}, (411456218, 257905): {}, (411456218, 258564): {}, (411456218, 258780): {}, (411456218, 259043): {}, (411456218, 259055): {}, (411456218, 261048): {}, (411456218, 261325): {}, (411456218, 261889): {}, .... (484619125, 46274059): {}, (484619125, 46274124): {}}`

Every item appears to be a 2-item tuple key and empty value; no apparent hierarchical structure.

Reply:
Ok, I found the problem. `hierarchy()` had

        selected_concept_ids = get_concept_set_member_ids(codeset_ids)

which returns tuples of (codeset_id, concept_id), but it needed a flat list of concept_ids, so had to specify the single column:

        selected_concept_ids = get_concept_set_member_ids(codeset_ids, column='concept_id')

Reply:
@joeflack4, not sure if you'll see the above if I resolve the conversation, but I will anyway